safekeeper/src/wal_backup.rs

```diff
@@ -0,0 +1,218 @@
+            .unwrap()
+    };
+}
+
+const BACKUP_ELECTION_PATH: &str = "WAL_BACKUP";
+
+// TODO: there's need to track last upload time to avoid losing partially filled segments
+// Situation: compute wrote a log record and went idle (log was backed up to S3)
+// SK got replaced, compute work up wrote log recrods enough to fill up the segment, 
+// we are going to upload the segment but writes before replacement will be missing
+
+#[derive(Clone)]
+pub struct WalBackup {
+    
+    // TODO: Can't figure out how to properly store this
+    _remote_storage : &LocalFs,
+ 
+    conf: SafeKeeperConf,
+
+    backup_lsn : Lsn,
+
+    // etcd lease id, used for leader election
+    lease_id : Option<i64>,
+    timeline_id: ZTenantTimelineId, 
+}
+
+async fn detect_task(
+    mut backup: WalBackup,
+    conf: SafeKeeperConf,
+    mut segment_complete: Receiver<(Lsn, Lsn)>,
+    mut lsn_durable: Receiver<Lsn>, ) {
+
+    loop {
+        if let Some(lease_id) = backup.lease_id {
+            broker::become_leader(lease_id, BACKUP_ELECTION_PATH.to_string(), &backup.timeline_id, conf.clone()).await.ok();
+        }
+
+        // TODO: can this result in us missing multiple events fired in a rapid succession?
+        while segment_complete.changed().await.is_ok() {
+            let segment = *segment_complete.borrow();
+            let mut commit_lsn = Lsn(0);
+            warn!("Woken Up for segment backup start {}, end {}", segment.0, segment.1);
+
+            // TODO: Make this spin optional
+            // if ! segment.1 is durable {
+            while lsn_durable.changed().await.is_ok() {
+                commit_lsn = *lsn_durable.borrow();
+
+                if segment.1 <= commit_lsn {
+                    warn!("Woken up for lsn {} is durable, not waiting anymore", commit_lsn);
+
+                    break;
+                }
+
+                warn!("Woken up for lsn {} is durable, will retry waiting", commit_lsn);
+            }
+
+            // This line assumes the previous task should have updated backup LSN correctly
+            backup.backup_segments(backup.backup_lsn, commit_lsn).await.ok();
+        }
+    }
+}
+
+
+#[allow(dead_code)]
+impl WalBackup {
+    pub fn create(remote_storage: LocalFs, conf: &SafeKeeperConf, timeline_id: &ZTenantTimelineId, segment_complete: Receiver<(Lsn, Lsn)>, lsn_durable: Receiver<Lsn>) -> Self {
+        warn!("Backup service is created");
+
+        let lease =
+        if conf.broker_endpoints.is_some() {
+            Some(broker::get_lease(conf))
+        } else {
+            None
+        };
+
+        let backup = Self {_remote_storage : &remote_storage, backup_lsn : Lsn::MAX , lease_id: lease, timeline_id: *timeline_id, conf: conf};
+
+        if let Some(lease_id) = lease {
+            // TODO: antons how to do it in this function???
+            // ensure!( conf.broker_endpoints.is_none(););
+
+            BACKUP_RUNTIME.spawn(broker::lease_keep_alive(lease_id, conf.clone()));
```

Comment:
We would need a cancellation mechanism here, similar to what's in the [similar place](https://github.com/neondatabase/neon/blob/778744d35ca4ff57237a6ef5b4323084797de9bd/pageserver/src/remote_storage/storage_sync.rs#L546) of pageserver.

What's worse is the fact that every `BACKUP_RUNTIME.spawn` returns a join handle, but we don't use that to wait for the task's completion, rather leaving it spinning in the runtime implicitly.

This means that the futures will be abruptly cancelled when the runtime is dropped only, abruptly terminating the upload streaming with that drop.
Should not be a big issue with S3 guarantees, but feels rather odd to see in general.

Reply:
Actually that's a bit scary and is valid for pagesever, too, if for some reason the uploader fails (panics etc.) but lease keepalives are still sent no one will be able to step in and replace the failed uploader

Reply:
agree, the cancellation story isn't sound, at the same time backup runtime is supposed to live for the whole duration of SK process life (new implementation makes it lazily initialized).
I'd like to address cancellation (incl. timeline drop) and SK replacement handling separately with the addition of time-based upload (in addition to LSN range). This would address keep-alive independence.

Reply:
Should we consider two different cancellation parts?

* "reasonable" bits such as reacting to SIGINT, being able to "shut down a tenant", etc. Not sure how it all is relevant for safekeeper, but we seem to not to have any means to support this in the new code
* sound cancellation with handling all edge cases, panics, etc.

The latter is out of scope for the current early version for sure, but what about regular termination commands?